hazelcast-hibernate5/src/test/java/com/hazelcast/hibernate/UpdateTimestampCacheTest.java

```diff
@@ -0,0 +1,138 @@
+        deleteDummyEntity(sf2, 1);
+
+        // Wait until invalidation messages are sent between sf and sf2
+        await().atMost(5, SECONDS).until(() -> isTimestampCacheUpToDate(sf) && isTimestampCacheUpToDate(sf2));
+
+        // Expect the cached query on sf is invalidated and the new one is inserted
+        executeQueryAndVerifyCacheStats(sf, entityCount - 2, 0, 2, 2);
+
+        // Expect the cached query on sf2 is invalidated and the new one is inserted
+        executeQueryAndVerifyCacheStats(sf2, entityCount - 2, 0, 2, 2);
+
+        // Verify sf query cache is not blocked
+        executeQueryAndVerifyCacheStats(sf, entityCount - 2, 1, 2, 2);
+
+        // Verify sf2 query cache is not blocked
+        executeQueryAndVerifyCacheStats(sf2, entityCount - 2, 1, 2, 2);
+    }
+
+    /**
+     * Runs #executeCacheableQuery on the factory and verifies the query result
+     * as well as the query cache stats after the query is executed.
+     *
+     * @param factory  factory to run the query
+     * @param expectedEntryCount  expected number of entries returned by the query
+     * @param expectedCacheHits  expected number of QueryCacheHitCount
+     * @param expectedCacheMisses  expected number of QueryCacheMissCount
+     * @param expectedCachePuts  expected number of QueryCachePutCount
+     */
+    private void executeQueryAndVerifyCacheStats(SessionFactory factory, int expectedEntryCount,
+                                                 long expectedCacheHits, long expectedCacheMisses,
+                                                 long expectedCachePuts) {
+        List<DummyEntity> list = executeCacheableQuery(factory);
+        assertEquals(expectedEntryCount, list.size());
+        assertEquals(expectedCacheHits, factory.getStatistics().getQueryCacheHitCount());
+        assertEquals(expectedCacheMisses, factory.getStatistics().getQueryCacheMissCount());
+        assertEquals(expectedCachePuts, factory.getStatistics().getQueryCachePutCount());
+    }
+
+    /**
+     * Executes a query to select all DummyEntity entries and caches the result.
+     *
+     * @param sessionFactory  factory to run the query
+     * @return  query result
+     */
+    private List<DummyEntity> executeCacheableQuery(SessionFactory sessionFactory) {
+        try (Session session = sessionFactory.openSession()) {
+            Query query = session.createQuery("from " + DummyEntity.class.getName());
+            query.setCacheable(true);
+            return query.list();
+        }
+    }
+
+    private boolean isTimestampCacheUpToDate(SessionFactory sessionFactory) {
+        try (Session session = sessionFactory.openSession()) {
+            UpdateTimestampsCache timestampsCache = ((SessionFactoryImpl) sessionFactory).getUpdateTimestampsCache();
+            return timestampsCache.isUpToDate(new HashSet<>(Arrays.asList("dummy_entities")),
+                    timestampsCache.getRegion().nextTimestamp(), (SessionImpl) session);
```

Comment:
otherwise `UpdateTimestampsCache#isUpToDate` returns true even if the cache is empty.
```suggestion
            return timestampsCache.getRegion().get((SessionImpl) session, "dummy_entities") != null &&
                    timestampsCache.isUpToDate(new HashSet<>(Arrays.asList("dummy_entities")),
                            timestampsCache.getRegion().nextTimestamp(), (SessionImpl) session);
```